### PR TITLE
Implement ListStatus RPC from client to Worker in Dora

### PR DESCRIPTION
**Pull request by huanghua78**
This first version of ListStatus from client to Dora worker.
More items are needed:
divide big directory contents into batch limited by max batch size.

### What changes are proposed in this pull request?

Dora client sends ListStatus() request to Dora Worker, and Worker sends ListStatus()
to UFS, and then the results are propagated back.

### Why are the changes needed?

This is the first step to do worker side ListStatus() cache.

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

**Comment by huanghua78** [comment]
Do we need `iterateStatus()`?
Do we need `listStatusPartial()`?

File: core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 1]
```diff
@@ -100,6 +103,7 @@ public class PagedDoraWorker extends AbstractWorker implements DoraWorker {
   private WorkerNetAddress mAddress;
 
   private RocksDBDoraMetaStore mMetaStore;
+  private UnderFileSystem mUfs;
```
Comment (LuQQiu) [inline core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:106]: final

File: core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -122,7 +126,7 @@ public PagedDoraWorker(AtomicReference<Long> workerId, AlluxioConfiguration conf
         .build(new CacheLoader<String, DoraMeta.FileStatus>() {
           @Override
           public DoraMeta.FileStatus load(String path) throws IOException {
-            UfsStatus status = ufs.getStatus(path);
+            UfsStatus status = mUfs.getStatus(path);
             DoraMeta.FileStatus fs = buildFileStatusFromUfsStatus(status, path);
             return fs;
```
Comment (LuQQiu) [inline core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:131]: inline? 
return buildFileStatusFromUfsStatus(mUfs.getStatus(path), path);

File: core/common/src/main/java/alluxio/worker/dora/DoraWorker.java [hunk 3]
```diff
@@ -33,6 +35,15 @@ public interface DoraWorker extends DataWorker, SessionCleanable {
...
   FileInfo getFileInfo(String fileId, GetStatusPOptions options) throws IOException;
 
+  /**
+   * List status from Under File System.
+   * @param path
+   * @param options
+   * @return list of status
+   * @throws IOException
+   */
+  UfsStatus[] listStatus(String path, ListOptions options) throws IOException;
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/dora/DoraWorker.java:45]: any reason that `FileInfo getFileInfo` but listStatus returns `UfsStatus[] listStatus`, ideally would want dora worker to return to the same kind of information, but this may be more a cleanup work, to cleanup the complicated and memory heavy data structures in Alluxio

**Review by LuQQiu — COMMENTED** [review]
Overall LGTM, leave some comments

File: core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:131]: I am not so familiar with the Java programming language.
IMHO, making it inline does not improve performance, nor less GC.
Then, why do I write the code like this?
Because when I do live debugging, it is easier for me to inspect the value of every variable and the return value from a function.

Maybe we can discuss this programming convention.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 4]
```diff
@@ -101,4 +106,35 @@ public void getStatus(GetStatusPRequest request,
...
+  @Override
+  public void listStatus(ListStatusPRequest request,
+                         StreamObserver<ListStatusPResponse> responseObserver) {
+    LOG.debug("listStatus is called for {}", request.getPath());
+
+    try {
+      ListOptions options = ListOptions.defaults().setRecursive(
+          request.getOptions().hasRecursive() ? request.getOptions().getRecursive() : false);
+      UfsStatus[] statuses;
+      statuses = mWorker.listStatus(request.getPath(), options);
```
[the remark this answers is not in the document]
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:119]: Yes. Fixed.

File: core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 1 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:106]: Yes. Fixed.

File: core/common/src/main/java/alluxio/worker/dora/DoraWorker.java (hunk 3 above)
Comment (huanghua78) [reply core/common/src/main/java/alluxio/worker/dora/DoraWorker.java:45]: Yes, good idea.
It's better to refine the `listStatus` to be `FileInfo[] listStatus()`.
I would handle this as a TODO.

**Comment by huanghua78** [comment]
> @huanghua78 for FUSE case, currently i haven't figured out the efficient way to use `listStatusPartial` but it's possible to use `listStatusPartial` to largely improve the performance and resource consumption on FUSE side.
> @tcrain FYI

That would be a performance optimization.
Let's do it step by step.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 5]
```diff
@@ -101,4 +106,35 @@ public void getStatus(GetStatusPRequest request,
...
+
+      ListStatusPResponse.Builder builder = ListStatusPResponse.newBuilder();
+
+      for (int i = 0; i < statuses.length; i++) {
+        UfsStatus status = statuses[i];
+        String ufsFullPath = status.getName();
+        alluxio.grpc.FileInfo fi =
+            ((PagedDoraWorker) mWorker).buildFileInfoFromUfsStatus(status, ufsFullPath);
```
[the remark this answers is not in the document]
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:127]: The same as I explain here: https://github.com/Alluxio/alluxio/pull/17043#discussion_r1130946638

**Review by beinan — APPROVED** [review]
lgtm

File: core/transport/src/main/proto/grpc/block_worker.proto [hunk 6]
```diff
@@ -16,6 +16,13 @@ service BlockWorker {
    * Returns the status of the file or directory.
    */
   rpc GetStatus (grpc.file.GetStatusPRequest) returns (grpc.file.GetStatusPResponse);
+  /**
+   * If the path points to a file, the method returns a singleton with its file information.
+   * If the path points to a directory, the method returns a list with file information for the
+   * directory contents.
+   */
+  rpc ListStatus(grpc.file.ListStatusPRequest) returns (stream grpc.file.ListStatusPResponse);
```
Comment (jja725) [inline core/transport/src/main/proto/grpc/block_worker.proto:24]: any reason we need a stream of responses that already contain a list?

Comment (huanghua78) [reply core/transport/src/main/proto/grpc/block_worker.proto:24]: Please refer to the file: core/transport/src/main/proto/grpc/file_system_master.proto

My understanding is:
a stream is used to transfer large chunks of data, e.g. for read/write, the data of a read response or a write request may exceed the limit of a single RPG normal message.

The same for a ListStatus() response. The response of the ListStatus() of a large dir with a lot of entries will too large to be fixed in a normal RPC message.
So, stream must be used to transfer the result.

Comment (jja725) [reply core/transport/src/main/proto/grpc/block_worker.proto:24]: I see, that make sense

**Review by jja725 — APPROVED** [review]

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 7]
```diff
@@ -101,4 +106,34 @@ public void getStatus(GetStatusPRequest request,
...
+
+  @Override
+  public void listStatus(ListStatusPRequest request,
+                         StreamObserver<ListStatusPResponse> responseObserver) {
+    LOG.debug("listStatus is called for {}", request.getPath());
+
+    try {
+      ListOptions options = ListOptions.defaults().setRecursive(
+          request.getOptions().hasRecursive() ? request.getOptions().getRecursive() : false);
+      UfsStatus[] statuses = mWorker.listStatus(request.getPath(), options);
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: Do we do chunked / paginated status listing? If the request is recursive and the directories structure is deep, this call can be expensive.

File: core/transport/src/main/proto/grpc/table/layout/hive/hive.proto [hunk 8]
```diff
@@ -6,7 +6,6 @@ option java_outer_classname = "HiveLayoutProto";
 
 package alluxio.grpc.table.layout;
 
-import "grpc/common.proto";
```
Comment (dbw9580) [inline core/transport/src/main/proto/grpc/table/layout/hive/hive.proto:9]: why this change?

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java [hunk 9]
```diff
@@ -172,4 +174,11 @@ StreamObserver<OpenLocalBlockRequest> openLocalBlock(
    * @return status
    */
   GetStatusPResponse getStatus(GetStatusPRequest request);
+
+  /**
+   * List status from Worker.
+   * @param request
+   * @return List of Status
+   */
+  java.util.Iterator<ListStatusPResponse> listStatus(ListStatusPRequest request);
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java:183]: Add an import for `java.util.Iterator`

**Review by dbw9580 — APPROVED** [review]
Mostly LGTM

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 7 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: That would be an Optimization for the next step.
I have explained this in the commit message:

`This first version of ListStatus from client to Dora worker.
More items are needed:
divide big directory contents into batch limited by max batch size.`

File: core/transport/src/main/proto/grpc/table/layout/hive/hive.proto (hunk 8 above)
Comment (huanghua78) [reply core/transport/src/main/proto/grpc/table/layout/hive/hive.proto:9]: Because when I tried to build the source, it complains some error like:
duplicated definition ....

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java (hunk 9 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java]: Fixed.

importing packages, or using the full package names, are the two ways to reference a package outside of this package.
Do we have a coding style requirement to "import a package" instead of "use full package name"?

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 7 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: I checked again the interfaces of `UnderFileSystem` which is used here.
There is no batching support in `UnderFileSystem`.

Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: But, yes, we can do this batching support in the service handler.
I will implement this later.

Comment (jja725) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: Support from ufs is planned. Haven't got a timeline yet. Metadata sync redesign also needs paginated list and we can probably get help from @tcrain or @elega

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java (hunk 9 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java]: I think it's "always import the symbol except when it's ambiguous. Then use the fully qualified name." Like what we do with `alluxio.grpc.FileInfo` and `alluxio.wire.FileInfo`.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 7 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:118]: Thank you, @jja725 .
Let's do this optimization when the UFS is ready.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java (hunk 9 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerClient.java]: Got it.

**Review by LuQQiu — APPROVED** [review]
LGTM

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please
